### PR TITLE
Improvements to prevent data loss in DynamoDB source

### LINKED ISSUE #3623: Retry when DynamoDB source fails to initialize due to permission issue
**Is your feature request related to a problem? Please describe.**
Currently, if DDB source role does not have `dynamodb:DescribeTable` permission, the pipeline will shut down during initialization. We can catch this as a retryable error like we did for Opensearch sink to allow user to correct permissions on the role.

The permission error is like this:
```
[dynamodb-pipeline-sink-worker-2-thread-1] ERROR org.opensearch.dataprepper.pipeline.common.PipelineThreadPoolExecutor - Pipeline [dynamodb-pipeline] process worker encountered a fatal exception, cannot proceed further
java.util.concurrent.ExecutionException: software.amazon.awssdk.services.dynamodb.model.DynamoDbException: User: arn:aws:sts::123456789012:assumed-role/pipelineRole/Data-Prepper-0233fade-6af1-4da2-a36b-923d9802c8cb is not authorized to perform: dynamodb:DescribeTable on resource: arn:aws:dynamodb:us-east-1:123456789012:table/test-source-table because no identity-based policy allows the dynamodb:DescribeTable action (Service: DynamoDb, Status Code: 400, Request ID: 839GJP2AL1PSOC8225BKVCULM3VV4KQNSO5AEMVJF66Q9ASUAAJG)
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[?:?]
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191) ~[?:?]
	at org.opensearch.dataprepper.pipeline.common.PipelineThreadPoolExecutor.afterExecute(PipelineThreadPoolExecutor.java:70) ~[data-prepper-core-2.6.0-SNAPSHOT.jar:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1129) ~[?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.base/java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: software.amazon.awssdk.services.dynamodb.model.DynamoDbException: User: arn:aws:sts::193777858833:assumed-role/pipelineRole/Data-Prepper-0233fade-6af1-4da2-a36b-923d9802c8cb is not authorized to perform: dynamodb:DescribeTable on resource: arn:aws:dynamodb:us-east-1:193777858833:table/gameday-test-source-table because no identity-based policy allows the dynamodb:DescribeTable action (Service: DynamoDb, Status Code: 400, Request ID: 839GJP2AL1PSOC8225BKVCULM3VV4KQNSO5AEMVJF66Q9ASUAAJG)
	at software.amazon.awssdk.core.internal.http.CombinedResponseHandler.handleErrorResponse(CombinedResponseHandler.java:125) ~[sdk-core-2.20.67.jar:?]
	...
	at software.amazon.awssdk.awscore.client.handler.AwsSyncClientHandler.execute(AwsSyncClientHandler.java:56) ~[aws-core-2.20.67.jar:?]
	at software.amazon.awssdk.services.dynamodb.DefaultDynamoDbClient.describeTable(DefaultDynamoDbClient.java:2344) ~[dynamodb-2.20.67.jar:?]
	at org.opensearch.dataprepper.plugins.source.dynamodb.DynamoDBService.getTableInfo(DynamoDBService.java:271) ~[dynamodb-source-2.6.0-SNAPSHOT.jar:?]
...
```

**Describe the solution you'd like**

**Describe alternatives you've considered (Optional)**

**Additional context**

### PR DESCRIPTION
**Pull request by daixba**
### Description
Some improvements made to DynamoDB source, mainly just to prevent data loss for some edge cases, changes include:
1.  Periodic Shard Discovery process used to find child shards for completed parents (Main change).
2.  Remove the support of TRIM_HORIZON for stream processing (Potential data loss if shard get expired)
3. Add a 5 minutes overlap between the export time and the event time
4. Always traverse from root shards instead of current active shards to ensure no missing shards

There are some other changes, such as more exceptional handling on initialization process which should fix issue #3623 
 
### Issues Resolved
Resolved #3623 
 
### Check List
- [x] New functionality includes testing.
- [x] New functionality has a documentation issue. Please link to it in this PR.
  - [x] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/configuration/StreamStartPosition.java [hunk 1]
```diff
@@ -5,7 +5,13 @@
...
 
+
+/**
+ * Always default to LATEST.
+ * Support of TRIM_HORIZON is taken out,
+ * that is due to a concern raised recently about the root shards may be expired while processing.
+ */
 public enum StreamStartPosition {
-    TRIM_HORIZON,
+    // TRIM_HORIZON,
```
Comment (dlvenable) [inline data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/configuration/StreamStartPosition.java:15]: Please remove the commented out code entirely.

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardConsumer.java [hunk 2]
```diff
@@ -183,9 +197,14 @@ public ShardConsumer build() {
     @Override
     public void run() {
         LOG.debug("Shard Consumer start to run...");
+        // Check should skip processing or not.
+        if (shouldSkip()) return;
 
         long lastCheckpointTime = System.currentTimeMillis();
         String sequenceNumber = "";
+        int interval;
+//        GetRecordsResponse response;
```
Comment (dlvenable) [inline data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardConsumer.java:206]: Please remove this commented-out line.

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardConsumerFactory.java [hunk 3]
```diff
@@ -84,10 +88,13 @@ public Runnable createConsumer(final StreamPartition streamPartition,
         String tableArn = getTableArn(streamPartition.getStreamArn());
         TableInfo tableInfo = getTableInfo(tableArn);
 
+        LOG.debug("Create shard consumer for {} with shardIter {}", streamPartition.getShardId(), shardIterator);
```
Comment (dlvenable) [inline data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardConsumerFactory.java:91]: I think it would be better to combine these log lines. Otherwise they will look unrelated when reviewing logs.

```
 LOG.debug("Create shard consumer for {} with shardIterator {} and lastShardIterator",
  streamPartition.getShardId(), shardIterator, lastShardIterator);
```

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardManager.java [hunk 4]
```diff
@@ -122,11 +124,12 @@ public String getShardIterator(String streamArn, String shardId, String sequence
         GetShardIteratorRequest getShardIteratorRequest;
 
         if (sequenceNumber != null && !sequenceNumber.isEmpty()) {
-            LOG.debug("Get Shard Iterator after " + sequenceNumber);
+            LOG.debug("Get Shard Iterator at " + sequenceNumber);
```
Comment (dlvenable) [inline data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardManager.java:127]: Please use SLF4J interpolation:

```
LOG.debug("Get Shard Iterator at {}", sequenceNumber);
```

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardConsumer.java [hunk 5]
```diff
@@ -298,6 +319,34 @@ private void waitForExport() {
...
 
+    /**
+     * Only to skip processing when below two conditions are met.
+     * - Last Shard Iterator is provided (Shard with ending sequence number)
+     * - Last Event Timestamp is later than start time or No Last Event Timestamp (empty shard)
+     */
+    private boolean shouldSkip() {
+        // Do skip check
+        if (lastShardIterator != null && !lastShardIterator.isEmpty()) {
+            GetRecordsResponse response = callGetRecords(lastShardIterator);
```
Comment (dlvenable) [inline data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardConsumer.java:330]: It appears that each iteration now requires two requests to `GetRecordsRequest` - one for the last shard iterator and another for the `shardIterator`. Is there a better way to avoid making both of these requests?

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardConsumer.java (hunk 5 above)
Comment (daixba) [reply data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardConsumer.java:329]: A shard iterator is used to get changes event (via GetRecords api) from stream, now to prevent missing shards, we will also process closed shards (shard with ending sequence number), only for closed shard, we will check the last record in the shard and see if the event time is after than the export time, if not, simply completed the shard. Otherwise, just use the normal shard iterator to get records for processing.

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardConsumerFactory.java (hunk 3 above)
Comment (daixba) [reply data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardConsumerFactory.java:94]: I agree. Only a shard iterator is too long (almost 1000 characters), so it's better to keep them in different lines.

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/DynamoDBService.java [hunk 6]
```diff
@@ -180,24 +177,14 @@ public void init() {
...
-                } else {
-                    shardIds = shardManager.getRootShardIds(tableInfo.getMetadata().getStreamArn());
-                    shardIds.forEach(shardId -> {
-                        createStreamPartition(tableInfo.getMetadata().getStreamArn(), shardId, null, false);
-                    });
-                }
+                // TODO: Revisit the use of start position.
+                // The behaviour is same for all cases regardless the configuration provided.
+                // Only process event from current date time but still traverse the shards from the beginning.
+                List<Shard> shards = shardManager.getRootShards(tableInfo.getMetadata().getStreamArn());
```
[the remark this answers is not in the document]
Comment (daixba) [reply data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/DynamoDBService.java]: Yes, root shards are the first parent.

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardConsumerFactory.java (hunk 3 above)
Comment (dlvenable) [reply data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardConsumerFactory.java:94]: That makes sense, so yes, let's keep it this way.

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/sourcecoordination/enhanced/EnhancedLeaseBasedSourceCoordinator.java [hunk 7]
```diff
@@ -125,6 +127,24 @@ public Optional<EnhancedSourcePartition> acquireAvailablePartition(String partit
         return Optional.of(partitionFactory.apply(sourceItem.get()));
     }
 
+    @Override
+    public List<EnhancedSourcePartition> queryCompletedPartitions(String partitionType, Instant fromCompletionTime) {
+        Objects.nonNull(partitionType);
```
Comment (dlvenable) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/sourcecoordination/enhanced/EnhancedLeaseBasedSourceCoordinator.java:132]: This statement doesn't do anything. I think you mean:

```
Objects.requireNonNull(partitionType);
```

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/leader/LeaderScheduler.java [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+                    // and then find children for all completed shards in the last 1 day (maximum time)
+                    // A shard will be expired after 24 hours.
+                    discoveryAndCreateChildrenPartitions();
+                    // Extend the timeout
+                    coordinator.saveProgressStateForPartition(leaderPartition, Duration.ofMinutes(DEFAULT_EXTEND_LEASE_MINUTES));
+
+                }
+
+            } catch (Exception e) {
+                e.printStackTrace();
```
Comment (dlvenable) [inline data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/leader/LeaderScheduler.java:105]: Please handle this differently. Either re-throw or log it using SLF4J.

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/leader/LeaderScheduler.java [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+                        shardId -> {
+                            if (!completedShardIds.contains(shardId)) {
+                                createChildStreamPartition(streamPartition, shardId);
+                            }
+                        }
+                );
+            }
+        });
+        long endTime = System.currentTimeMillis();
+        LOG.info("Shard discovery process is completed, it took {} milliseconds in total", endTime - startTime);
```
Comment (dlvenable) [inline data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/leader/LeaderScheduler.java:206]: Should we report this as a metric instead of log it? Micrometer has a nice `timer()` that you can use.

https://github.com/opensearch-project/data-prepper/blob/d26748680468aa50bf1bf93c83cc005ae423a2e0/data-prepper-api/src/main/java/org/opensearch/dataprepper/metrics/PluginMetrics.java#L62-L64

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardConsumer.java (hunk 5 above)
Comment (graytaylor0) [reply data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardConsumer.java:329]: I think the question is why we call `callGetRecords` in the `shouldSkip` check, and then call it again right after. Why not call it one time and reuse that response to check the skip and then you don't need to call it again on line 222

File: data-prepper-plugins/in-memory-source-coordination-store/src/main/java/org/opensearch/dataprepper/plugins/sourcecoordinator/inmemory/InMemorySourceCoordinationStore.java [hunk 10]
```diff
@@ -56,6 +58,11 @@ public Optional<SourcePartitionStoreItem> getSourcePartitionItem(final String so
         }
     }
 
+    @Override
+    public List<SourcePartitionStoreItem> querySourcePartitionItemsByStatus(final String sourceIdentifier, final String sourcePartitionKey, final SourcePartitionStatus sourcePartitionStatus, final String startPartitionPriority) {
+        throw new RuntimeException("querySourcePartitionItemsByStatus is currently not supported in In Memory Store");
```
Comment (graytaylor0) [inline data-prepper-plugins/in-memory-source-coordination-store/src/main/java/org/opensearch/dataprepper/plugins/sourcecoordinator/inmemory/InMemorySourceCoordinationStore.java:63]: You can alternatively throw an `UnsupportedOperationException`

**Review by graytaylor0 — COMMENTED** [review]
Had some small comments. You also need to rebase, and other than that looks good

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardConsumer.java (hunk 5 above)
Comment (daixba) [reply data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardConsumer.java:329]: That was actually I did at the beginning,  passed something like `lastShardIterator == null? shardIterator: lastShardIterator` to the getRecords request, then set `lastShardIterator = null` if the should process to go to the normal process.  But it doesn't look tidy to me though there is only one GetRecords call in the code, it's just to check `if lastShardIterator == null` everywhere and differnet logic applies.   So I completely add a separate section for lastShardIterator which can be taken out in the future, there is another option that we don't need to process closed shard probably is something need to try in the future.

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/leader/LeaderScheduler.java (hunk 9 above)
Comment (daixba) [reply data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/leader/LeaderScheduler.java:206]: I put this as info log because I can't test the performance in my local, to call DDB api from my local, the latency is too high. So this process may take less than 1 second in OSIS v.s. many seconds in my local.

We will test the code, and if the performance is ok, we can change this to debug.  At least right now, this information is still for testing purpose, not sure if it has a production value, but can add it as a metric in the future.

File: data-prepper-plugins/in-memory-source-coordination-store/src/main/java/org/opensearch/dataprepper/plugins/sourcecoordinator/inmemory/InMemorySourceCoordinationStore.java (hunk 10 above)
Comment (daixba) [reply data-prepper-plugins/in-memory-source-coordination-store/src/main/java/org/opensearch/dataprepper/plugins/sourcecoordinator/inmemory/InMemorySourceCoordinationStore.java:63]: Will change.

File: data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/leader/LeaderScheduler.java (hunk 8 above)
Comment (daixba) [reply data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/leader/LeaderScheduler.java:105]: Sure, will make the change.

File: data-prepper-plugins/dynamodb-source-coordination-store/src/main/java/org/opensearch/dataprepper/plugins/sourcecoordinator/dynamodb/DynamoDbClientWrapper.java [hunk 11]
```diff
@@ -252,12 +255,37 @@ public Optional<SourcePartitionStoreItem> getAvailablePartition(final String own
...
     }
+
+
+    public List<SourcePartitionStoreItem> queryPartitionsByStatus(final String sourceStatusCombinationKey, final String partitionPriority) {
+        List<SourcePartitionStoreItem> result = new ArrayList<>();
+        try {
+            final DynamoDbIndex<DynamoDbSourcePartitionItem> sourceStatusIndex = table.index(SOURCE_STATUS_COMBINATION_KEY_GLOBAL_SECONDARY_INDEX);
+            final QueryEnhancedRequest queryEnhancedRequest = QueryEnhancedRequest.builder()
+                    .limit(DEFAULT_QUERY_LIMIT)
+                    .queryConditional(QueryConditional.sortGreaterThan(Key.builder().partitionValue(sourceStatusCombinationKey).sortValue(partitionPriority).build()))
```
[the remark this answers is not in the document]
Comment (daixba) [reply data-prepper-plugins/dynamodb-source-coordination-store/src/main/java/org/opensearch/dataprepper/plugins/sourcecoordinator/dynamodb/DynamoDbClientWrapper.java:273]: I think it doesn't matter here, we don't need to do exactly 24 hours, which is the maximum time required, actually for those completed in exactly 24 hours ago should already be expired.

**Review by graytaylor0 — APPROVED** [review]

**Review by kkondaka — APPROVED** [review]
